// carbon-components-angular/esm2020/table/header/table-header-title.directive.mjs
import { Directive, HostBinding, Input } from "@angular/core";
import * as i0 from "@angular/core";
export class TableHeaderTitle {
    constructor() {
        this.id = `table-title-${TableHeaderTitle.counter++}`;
        this.titleClass = true;
    }
}
TableHeaderTitle.counter = 0;
TableHeaderTitle.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeaderTitle, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TableHeaderTitle.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: TableHeaderTitle, selector: "[cdsTableHeaderTitle], [ibmTableHeaderTitle]", inputs: { id: "id" }, host: { properties: { "attr.id": "this.id", "class.cds--data-table-header__title": "this.titleClass" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeaderTitle, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsTableHeaderTitle], [ibmTableHeaderTitle]"
                }]
        }], propDecorators: { id: [{
                type: HostBinding,
                args: ["attr.id"]
            }, {
                type: Input
            }], titleClass: [{
                type: HostBinding,
                args: ["class.cds--data-table-header__title"]
            }] } });
