// carbon-components-angular/esm2020/table/header/table-header-description.directive.mjs
import { Directive, HostBinding, Input } from "@angular/core";
import * as i0 from "@angular/core";
export class TableHeaderDescription {
    constructor() {
        this.id = `table-description-${TableHeaderDescription.counter++}`;
        this.descriptionClass = true;
    }
}
TableHeaderDescription.counter = 0;
TableHeaderDescription.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeaderDescription, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TableHeaderDescription.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: TableHeaderDescription, selector: "[cdsTableHeaderDescription], [ibmTableHeaderDescription]", inputs: { id: "id" }, host: { properties: { "attr.id": "this.id", "class.cds--data-table-header__description": "this.descriptionClass" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeaderDescription, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsTableHeaderDescription], [ibmTableHeaderDescription]"
                }]
        }], propDecorators: { id: [{
                type: HostBinding,
                args: ["attr.id"]
            }, {
                type: Input
            }], descriptionClass: [{
                type: HostBinding,
                args: ["class.cds--data-table-header__description"]
            }] } });
