// carbon-components-angular/esm2020/table/header/table-header.component.mjs
import { Component, HostBinding } from "@angular/core";
import * as i0 from "@angular/core";
export class TableHeader {
    constructor() {
        this.headerClass = true;
        this.displayStyle = "block";
    }
}
TableHeader.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeader, deps: [], target: i0.ɵɵFactoryTarget.Component });
TableHeader.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: TableHeader, selector: "cds-table-header, ibm-table-header", host: { properties: { "class.cds--data-table-header": "this.headerClass", "style.display": "this.displayStyle" } }, ngImport: i0, template: `
		<ng-content></ng-content>
	`, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeader, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-table-header, ibm-table-header",
                    template: `
		<ng-content></ng-content>
	`
                }]
        }], propDecorators: { headerClass: [{
                type: HostBinding,
                args: ["class.cds--data-table-header"]
            }], displayStyle: [{
                type: HostBinding,
                args: ["style.display"]
            }] } });
