// carbon-components-angular/fesm2020/carbon-components-angular-forms.mjs
import * as i0 from '@angular/core';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { CheckboxModule } from 'carbon-components-angular/checkbox';
export { Checkbox, CheckboxModule, CheckboxState } from 'carbon-components-angular/checkbox';
import { ToggleModule } from 'carbon-components-angular/toggle';
export { Toggle, ToggleModule, ToggleState } from 'carbon-components-angular/toggle';
import { RadioModule } from 'carbon-components-angular/radio';
export { Radio, RadioChange, RadioGroup, RadioModule } from 'carbon-components-angular/radio';
import { InputModule } from 'carbon-components-angular/input';
export { InputModule, Label, TextArea, TextInput, TextInputLabelComponent, TextareaLabelComponent } from 'carbon-components-angular/input';
import { ButtonModule } from 'carbon-components-angular/button';
export { Button, ButtonModule, ButtonSet, IconButton } from 'carbon-components-angular/button';

// modules
class NFormsModule {
}
NFormsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: NFormsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NFormsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: NFormsModule, imports: [CommonModule,
        FormsModule,
        CheckboxModule,
        ToggleModule,
        RadioModule,
        InputModule,
        ButtonModule], exports: [CheckboxModule,
        ToggleModule,
        RadioModule,
        InputModule,
        ButtonModule,
        InputModule] });
NFormsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: NFormsModule, imports: [CommonModule,
        FormsModule,
        CheckboxModule,
        ToggleModule,
        RadioModule,
        InputModule,
        ButtonModule, CheckboxModule,
        ToggleModule,
        RadioModule,
        InputModule,
        ButtonModule,
        InputModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: NFormsModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [
                        CheckboxModule,
                        ToggleModule,
                        RadioModule,
                        InputModule,
                        ButtonModule,
                        InputModule
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        CheckboxModule,
                        ToggleModule,
                        RadioModule,
                        InputModule,
                        ButtonModule
                    ]
                }]
        }] });

// export the forms module

/**
 * Generated bundle index. Do not edit.
 */

export { NFormsModule };
