// carbon-components-angular/esm2020/dropdown/dropdown.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { FormsModule } from "@angular/forms";
import { Dropdown } from "./dropdown.component";
import { DropdownList } from "./list/dropdown-list.component";
import { AbstractDropdownView } from "./abstract-dropdown-view.class";
import { ScrollableList } from "./scrollable-list.directive";
import { I18nModule } from "carbon-components-angular/i18n";
import { PlaceholderModule } from "carbon-components-angular/placeholder";
import { DropdownService } from "./dropdown.service";
import { UtilsModule } from "carbon-components-angular/utils";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class DropdownModule {
}
DropdownModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DropdownModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DropdownModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: DropdownModule, declarations: [Dropdown,
        DropdownList,
        ScrollableList,
        AbstractDropdownView], imports: [CommonModule,
        FormsModule,
        I18nModule,
        PlaceholderModule,
        UtilsModule,
        IconModule], exports: [Dropdown,
        DropdownList,
        ScrollableList,
        AbstractDropdownView] });
DropdownModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DropdownModule, providers: [DropdownService], imports: [CommonModule,
        FormsModule,
        I18nModule,
        PlaceholderModule,
        UtilsModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DropdownModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Dropdown,
                        DropdownList,
                        ScrollableList,
                        AbstractDropdownView
                    ],
                    exports: [
                        Dropdown,
                        DropdownList,
                        ScrollableList,
                        AbstractDropdownView
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        I18nModule,
                        PlaceholderModule,
                        UtilsModule,
                        IconModule
                    ],
                    providers: [DropdownService]
                }]
        }] });
