// carbon-components-angular/esm2020/datepicker-input/datepicker-input.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { DatePickerInput } from "./datepicker-input.component";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class DatePickerInputModule {
}
DatePickerInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DatePickerInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DatePickerInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: DatePickerInputModule, declarations: [DatePickerInput], imports: [CommonModule,
        IconModule], exports: [DatePickerInput] });
DatePickerInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DatePickerInputModule, imports: [CommonModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DatePickerInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        DatePickerInput
                    ],
                    exports: [
                        DatePickerInput
                    ],
                    imports: [
                        CommonModule,
                        IconModule
                    ]
                }]
        }] });
