// carbon-components-angular/esm2020/datepicker/datepicker.module.mjs
import { DatePickerInputModule } from "carbon-components-angular/datepicker-input";
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { DatePicker } from "./datepicker.component";
import { UtilsModule } from "carbon-components-angular/utils";
import { I18nModule } from "carbon-components-angular/i18n";
import * as i0 from "@angular/core";
export class DatePickerModule {
}
DatePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DatePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DatePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: DatePickerModule, declarations: [DatePicker], imports: [CommonModule,
        DatePickerInputModule,
        UtilsModule,
        I18nModule], exports: [DatePicker,
        DatePickerInputModule] });
DatePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DatePickerModule, imports: [CommonModule,
        DatePickerInputModule,
        UtilsModule,
        I18nModule, DatePickerInputModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DatePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        DatePicker
                    ],
                    exports: [
                        DatePicker,
                        DatePickerInputModule
                    ],
                    imports: [
                        CommonModule,
                        DatePickerInputModule,
                        UtilsModule,
                        I18nModule
                    ]
                }]
        }] });
