// carbon-components-angular/esm2020/toggle/toggle.module.mjs
// modules
import { NgModule } from "@angular/core";
import { FormsModule } from "@angular/forms";
import { CommonModule } from "@angular/common";
// imports
import { I18nModule } from "carbon-components-angular/i18n";
import { Toggle } from "./toggle.component";
import * as i0 from "@angular/core";
export class ToggleModule {
}
ToggleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ToggleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ToggleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ToggleModule, declarations: [Toggle], imports: [CommonModule,
        FormsModule,
        I18nModule], exports: [Toggle] });
ToggleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ToggleModule, imports: [CommonModule,
        FormsModule,
        I18nModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ToggleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Toggle
                    ],
                    exports: [
                        Toggle
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        I18nModule
                    ]
                }]
        }] });
