// carbon-components-angular/esm2020/layer/layer.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { LayerDirective } from "./layer.directive";
import * as i0 from "@angular/core";
export class LayerModule {
}
LayerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LayerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LayerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: LayerModule, declarations: [LayerDirective], imports: [CommonModule], exports: [LayerDirective] });
LayerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LayerModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LayerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LayerDirective],
                    exports: [LayerDirective],
                    imports: [CommonModule]
                }]
        }] });
