// carbon-components-angular/esm2020/inline-loading/inline-loading.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { InlineLoading } from "./inline-loading.component";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class InlineLoadingModule {
}
InlineLoadingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: InlineLoadingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
InlineLoadingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: InlineLoadingModule, declarations: [InlineLoading], imports: [CommonModule, IconModule], exports: [InlineLoading] });
InlineLoadingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: InlineLoadingModule, imports: [CommonModule, IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: InlineLoadingModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [InlineLoading],
                    exports: [InlineLoading],
                    imports: [CommonModule, IconModule]
                }]
        }] });
