// carbon-components-angular/esm2020/search/search.module.mjs
// modules
import { NgModule } from "@angular/core";
import { FormsModule } from "@angular/forms";
import { CommonModule } from "@angular/common";
// imports
import { I18nModule } from "carbon-components-angular/i18n";
import { Search } from "./search.component";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class SearchModule {
}
SearchModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SearchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SearchModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: SearchModule, declarations: [Search], imports: [FormsModule,
        CommonModule,
        I18nModule,
        IconModule], exports: [Search] });
SearchModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SearchModule, imports: [FormsModule,
        CommonModule,
        I18nModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SearchModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Search
                    ],
                    exports: [
                        Search
                    ],
                    imports: [
                        FormsModule,
                        CommonModule,
                        I18nModule,
                        IconModule
                    ]
                }]
        }] });
