// carbon-components-angular/esm2020/loading/loading.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { Loading } from "./loading.component";
import { I18nModule } from "carbon-components-angular/i18n";
import * as i0 from "@angular/core";
export class LoadingModule {
}
LoadingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LoadingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LoadingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: LoadingModule, declarations: [Loading], imports: [CommonModule, I18nModule], exports: [Loading] });
LoadingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LoadingModule, imports: [CommonModule, I18nModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LoadingModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [Loading],
                    exports: [Loading],
                    imports: [CommonModule, I18nModule]
                }]
        }] });
