// carbon-components-angular/esm2020/structured-list/list-column.component.mjs
import { Component, HostBinding, Input } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
/**
 * `ListColumn` represents a single column in a `StructuredList`.
 *
 * `ListColumn`s can be used in a `ListHeader` to specify the column headers, or in `ListRow`s to specify the column data.
 *
 * Example:
 * ```html
 * <cds-list-column nowrap="true">Column 1</cds-list-column>
 * ```
 */
export class ListColumn {
    constructor() {
        this.skeleton = false;
        this.role = "cell";
        this.isHeaderColumn = true;
        this.isBodyColumn = true;
        /**
         * Applies `white-space: nowrap` to the content of this `ListColumn`
         */
        this.nowrap = false;
    }
}
ListColumn.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ListColumn, deps: [], target: i0.ɵɵFactoryTarget.Component });
ListColumn.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: ListColumn, selector: "cds-list-column, ibm-list-column", inputs: { skeleton: "skeleton", nowrap: "nowrap" }, host: { properties: { "attr.role": "this.role", "class.cds--structured-list-th": "this.isHeaderColumn", "class.cds--structured-list-td": "this.isBodyColumn", "class.cds--structured-list-content--nowrap": "this.nowrap" } }, ngImport: i0, template: `
		<span *ngIf="skeleton"></span>
		<ng-content></ng-content>
	`, isInline: true, dependencies: [{ kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ListColumn, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-list-column, ibm-list-column",
                    template: `
		<span *ngIf="skeleton"></span>
		<ng-content></ng-content>
	`
                }]
        }], propDecorators: { skeleton: [{
                type: Input
            }], role: [{
                type: HostBinding,
                args: ["attr.role"]
            }], isHeaderColumn: [{
                type: HostBinding,
                args: ["class.cds--structured-list-th"]
            }], isBodyColumn: [{
                type: HostBinding,
                args: ["class.cds--structured-list-td"]
            }], nowrap: [{
                type: HostBinding,
                args: ["class.cds--structured-list-content--nowrap"]
            }, {
                type: Input
            }] } });
