// carbon-components-angular/esm2020/structured-list/structured-list.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { StructuredList } from "./structured-list.component";
import { ListRow } from "./list-row.component";
import { ListHeader } from "./list-header.component";
import { ListColumn } from "./list-column.component";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class StructuredListModule {
}
StructuredListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: StructuredListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StructuredListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: StructuredListModule, declarations: [StructuredList,
        ListRow,
        ListHeader,
        ListColumn], imports: [CommonModule,
        IconModule], exports: [StructuredList,
        ListRow,
        ListHeader,
        ListColumn] });
StructuredListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: StructuredListModule, imports: [CommonModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: StructuredListModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        StructuredList,
                        ListRow,
                        ListHeader,
                        ListColumn
                    ],
                    exports: [
                        StructuredList,
                        ListRow,
                        ListHeader,
                        ListColumn
                    ],
                    imports: [
                        CommonModule,
                        IconModule
                    ]
                }]
        }] });
