// carbon-components-angular/esm2020/modal/modal-content.directive.mjs
import { Directive, HostBinding, Input } from "@angular/core";
import * as i0 from "@angular/core";
export class ModalContent {
    constructor() {
        this.modalContentClass = true;
        /**
         * Provide whether the modal content has a form element.
         * If `true` is used here, non-form child content should have `cds--modal-content__regular-content` class.
         */
        this.hasForm = false;
    }
}
ModalContent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ModalContent, deps: [], target: i0.ɵɵFactoryTarget.Directive });
ModalContent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: ModalContent, selector: "[cdsModalContent], [ibmModalContent]", inputs: { hasForm: "hasForm" }, host: { properties: { "class.cds--modal-content": "this.modalContentClass", "class.cds--modal-content--with-form": "this.hasForm" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ModalContent, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsModalContent], [ibmModalContent]"
                }]
        }], propDecorators: { modalContentClass: [{
                type: HostBinding,
                args: ["class.cds--modal-content"]
            }], hasForm: [{
                type: HostBinding,
                args: ["class.cds--modal-content--with-form"]
            }, {
                type: Input
            }] } });
