// carbon-components-angular/esm2020/modal/modal.module.mjs
// modules
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
// imports
import { BaseModalService } from "./base-modal.service";
import { ModalService } from "./modal.service";
import { Modal } from "./modal.component";
import { ModalFooter } from "./modal-footer.component";
import { Overlay } from "./overlay.component";
import { ModalHeader } from "./modal-header.component";
import { AlertModal } from "./alert-modal.component";
import { ButtonModule } from "carbon-components-angular/forms";
import { I18nModule } from "carbon-components-angular/i18n";
import { PlaceholderModule } from "carbon-components-angular/placeholder";
import { ExperimentalModule } from "carbon-components-angular/experimental";
import { ModalContent } from "./modal-content.directive";
import { ModalContentText } from "./modal-content-text.directive";
import { ModalHeaderHeading } from "./modal-header-heading.directive";
import { ModalHeaderLabel } from "./modal-header-label.directive";
import { BaseModal } from "./base-modal.class";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class ModalModule {
}
ModalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ModalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ModalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ModalModule, declarations: [AlertModal,
        Modal,
        ModalHeader,
        ModalFooter,
        Overlay,
        ModalContent,
        ModalContentText,
        ModalHeaderHeading,
        ModalHeaderLabel,
        BaseModal], imports: [CommonModule,
        ButtonModule,
        I18nModule,
        PlaceholderModule,
        ExperimentalModule,
        IconModule], exports: [AlertModal,
        Modal,
        ModalHeader,
        ModalFooter,
        ModalContent,
        ModalContentText,
        ModalHeaderHeading,
        ModalHeaderLabel,
        BaseModal] });
ModalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ModalModule, providers: [BaseModalService, ModalService], imports: [CommonModule,
        ButtonModule,
        I18nModule,
        PlaceholderModule,
        ExperimentalModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ModalModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        AlertModal,
                        Modal,
                        ModalHeader,
                        ModalFooter,
                        Overlay,
                        ModalContent,
                        ModalContentText,
                        ModalHeaderHeading,
                        ModalHeaderLabel,
                        BaseModal
                    ],
                    exports: [
                        AlertModal,
                        Modal,
                        ModalHeader,
                        ModalFooter,
                        ModalContent,
                        ModalContentText,
                        ModalHeaderHeading,
                        ModalHeaderLabel,
                        BaseModal
                    ],
                    providers: [BaseModalService, ModalService],
                    imports: [
                        CommonModule,
                        ButtonModule,
                        I18nModule,
                        PlaceholderModule,
                        ExperimentalModule,
                        IconModule
                    ]
                }]
        }] });
