// carbon-components-angular/esm2020/select/select.module.mjs
// modules
import { NgModule } from "@angular/core";
import { FormsModule } from "@angular/forms";
import { CommonModule } from "@angular/common";
// imports
import { Select } from "./select.component";
import { Option } from "./option.directive";
import { OptGroup } from "./optgroup.directive";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class SelectModule {
}
SelectModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SelectModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SelectModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: SelectModule, declarations: [Select,
        Option,
        OptGroup], imports: [CommonModule,
        FormsModule,
        IconModule], exports: [Select,
        Option,
        OptGroup] });
SelectModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SelectModule, imports: [CommonModule,
        FormsModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SelectModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Select,
                        Option,
                        OptGroup
                    ],
                    exports: [
                        Select,
                        Option,
                        OptGroup
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        IconModule
                    ]
                }]
        }] });
