// carbon-components-angular/esm2020/table/toolbar/table-toolbar-content.component.mjs
import { Component, HostBinding } from "@angular/core";
import * as i0 from "@angular/core";
export class TableToolbarContent {
    constructor() {
        this.class = true;
    }
}
TableToolbarContent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableToolbarContent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TableToolbarContent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: TableToolbarContent, selector: "cds-table-toolbar-content, ibm-table-toolbar-content", host: { properties: { "class.cds--toolbar-content": "this.class" } }, ngImport: i0, template: `<ng-content></ng-content>`, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableToolbarContent, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-table-toolbar-content, ibm-table-toolbar-content",
                    template: `<ng-content></ng-content>`
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ["class.cds--toolbar-content"]
            }] } });
