// carbon-components-angular/esm2020/accordion/accordion.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { Accordion } from "./accordion.component";
import { AccordionItem } from "./accordion-item.component";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class AccordionModule {
}
AccordionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: AccordionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AccordionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: AccordionModule, declarations: [Accordion,
        AccordionItem], imports: [CommonModule,
        IconModule], exports: [Accordion,
        AccordionItem] });
AccordionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: AccordionModule, imports: [CommonModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: AccordionModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Accordion,
                        AccordionItem
                    ],
                    exports: [
                        Accordion,
                        AccordionItem
                    ],
                    imports: [
                        CommonModule,
                        IconModule
                    ]
                }]
        }] });
