// carbon-components-angular/esm2020/ui-shell/header/header-global.component.mjs
import { Component, HostBinding } from "@angular/core";
import * as i0 from "@angular/core";
/**
 * Container for `HeaderAction`s.
 */
export class HeaderGlobal {
    constructor() {
        this.hostClass = true;
    }
}
HeaderGlobal.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: HeaderGlobal, deps: [], target: i0.ɵɵFactoryTarget.Component });
HeaderGlobal.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: HeaderGlobal, selector: "cds-header-global, ibm-header-global", host: { properties: { "class.cds--header__global": "this.hostClass" } }, ngImport: i0, template: `
		<ng-content></ng-content>
	`, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: HeaderGlobal, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-header-global, ibm-header-global",
                    template: `
		<ng-content></ng-content>
	`
                }]
        }], propDecorators: { hostClass: [{
                type: HostBinding,
                args: ["class.cds--header__global"]
            }] } });
