// carbon-components-angular/esm2020/ui-shell/header/header.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ButtonModule } from "carbon-components-angular/button";
import { I18nModule } from "carbon-components-angular/i18n";
import { IconModule } from "carbon-components-angular/icon";
import { Header } from "./header.component";
import { HeaderItem } from "./header-item.component";
import { HeaderMenu } from "./header-menu.component";
import { HeaderNavigation } from "./header-navigation.component";
import { HeaderGlobal } from "./header-global.component";
import { HeaderAction } from "./header-action.component";
import { Hamburger } from "./hamburger.component";
import { RouterModule } from "@angular/router";
import * as i0 from "@angular/core";
export { Header, HeaderItem, HeaderMenu, HeaderNavigation, HeaderGlobal, HeaderAction, Hamburger };
export class HeaderModule {
}
HeaderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: HeaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
HeaderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: HeaderModule, declarations: [Header,
        HeaderItem,
        HeaderMenu,
        HeaderNavigation,
        HeaderGlobal,
        HeaderAction,
        Hamburger], imports: [CommonModule,
        ButtonModule,
        I18nModule,
        IconModule,
        RouterModule], exports: [Header,
        HeaderItem,
        HeaderMenu,
        HeaderNavigation,
        HeaderGlobal,
        HeaderAction,
        Hamburger] });
HeaderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: HeaderModule, imports: [CommonModule,
        ButtonModule,
        I18nModule,
        IconModule,
        RouterModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: HeaderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Header,
                        HeaderItem,
                        HeaderMenu,
                        HeaderNavigation,
                        HeaderGlobal,
                        HeaderAction,
                        Hamburger
                    ],
                    imports: [
                        CommonModule,
                        ButtonModule,
                        I18nModule,
                        IconModule,
                        RouterModule
                    ],
                    exports: [
                        Header,
                        HeaderItem,
                        HeaderMenu,
                        HeaderNavigation,
                        HeaderGlobal,
                        HeaderAction,
                        Hamburger
                    ]
                }]
        }] });
