// carbon-components-angular/esm2020/number-input/number.module.mjs
// modules
import { NgModule } from "@angular/core";
import { FormsModule } from "@angular/forms";
import { CommonModule } from "@angular/common";
// imports
import { Number } from "./number.component";
import { I18nModule } from "carbon-components-angular/i18n";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class NumberModule {
}
NumberModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: NumberModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NumberModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: NumberModule, declarations: [Number], imports: [FormsModule,
        CommonModule,
        I18nModule,
        IconModule], exports: [Number] });
NumberModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: NumberModule, imports: [FormsModule,
        CommonModule,
        I18nModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: NumberModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Number
                    ],
                    exports: [
                        Number
                    ],
                    imports: [
                        FormsModule,
                        CommonModule,
                        I18nModule,
                        IconModule
                    ]
                }]
        }] });
