// carbon-components-angular/esm2020/radio/radio.module.mjs
// modules
import { NgModule } from "@angular/core";
import { FormsModule } from "@angular/forms";
import { CommonModule } from "@angular/common";
// imports
import { Radio } from "./radio.component";
import { RadioGroup } from "./radio-group.component";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class RadioModule {
}
RadioModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: RadioModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RadioModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: RadioModule, declarations: [Radio,
        RadioGroup], imports: [CommonModule,
        FormsModule,
        IconModule], exports: [Radio,
        RadioGroup] });
RadioModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: RadioModule, imports: [CommonModule,
        FormsModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: RadioModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Radio,
                        RadioGroup
                    ],
                    exports: [
                        Radio,
                        RadioGroup
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        IconModule
                    ]
                }]
        }] });
