// carbon-components-angular/esm2020/table/head/table-head-expand.component.mjs
import { Component, HostBinding } from "@angular/core";
import * as i0 from "@angular/core";
export class TableHeadExpand {
    constructor() {
        this.hostClass = true;
    }
}
TableHeadExpand.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeadExpand, deps: [], target: i0.ɵɵFactoryTarget.Component });
TableHeadExpand.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: TableHeadExpand, selector: "[cdsTableHeadExpand], [ibmTableHeadExpand]", host: { properties: { "class.cds--table-expand": "this.hostClass" } }, ngImport: i0, template: `
		<ng-content></ng-content>
	`, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeadExpand, decorators: [{
            type: Component,
            args: [{
                    // tslint:disable-next-line: component-selector
                    selector: "[cdsTableHeadExpand], [ibmTableHeadExpand]",
                    template: `
		<ng-content></ng-content>
	`
                }]
        }], propDecorators: { hostClass: [{
                type: HostBinding,
                args: ["class.cds--table-expand"]
            }] } });
