// carbon-components-angular/esm2020/progress-bar/progress-bar.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ProgressBar } from "./progress-bar.component";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class ProgressBarModule {
}
ProgressBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ProgressBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ProgressBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ProgressBarModule, declarations: [ProgressBar], imports: [CommonModule,
        IconModule], exports: [ProgressBar] });
ProgressBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ProgressBarModule, imports: [CommonModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ProgressBarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ProgressBar
                    ],
                    exports: [
                        ProgressBar
                    ],
                    imports: [
                        CommonModule,
                        IconModule
                    ]
                }]
        }] });
