// carbon-components-angular/esm2020/popover/popover.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { PopoverContainer } from "./popover.directive";
import { PopoverContent } from "./popover-content.component";
import * as i0 from "@angular/core";
export class PopoverModule {
}
PopoverModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PopoverModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PopoverModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: PopoverModule, declarations: [PopoverContainer,
        PopoverContent], imports: [CommonModule], exports: [PopoverContainer,
        PopoverContent] });
PopoverModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PopoverModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PopoverModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        PopoverContainer,
                        PopoverContent
                    ],
                    exports: [
                        PopoverContainer,
                        PopoverContent
                    ],
                    imports: [CommonModule]
                }]
        }] });
