// carbon-components-angular/esm2020/popover/popover-content.component.mjs
import { Component, HostBinding } from "@angular/core";
import * as i0 from "@angular/core";
/**
 * [See demo](../../?path=/story/components-popover--basic)
 */
export class PopoverContent {
    constructor() {
        this.popoverClass = true;
    }
}
PopoverContent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PopoverContent, deps: [], target: i0.ɵɵFactoryTarget.Component });
PopoverContent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: PopoverContent, selector: "cds-popover-content, ibm-popover-content", host: { properties: { "class.cds--popover": "this.popoverClass" } }, ngImport: i0, template: `
		<span class="cds--popover-content">
			<ng-content></ng-content>
		</span>
		<span class="cds--popover-caret"></span>
	`, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PopoverContent, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-popover-content, ibm-popover-content",
                    template: `
		<span class="cds--popover-content">
			<ng-content></ng-content>
		</span>
		<span class="cds--popover-caret"></span>
	`
                }]
        }], propDecorators: { popoverClass: [{
                type: HostBinding,
                args: ["class.cds--popover"]
            }] } });
