// carbon-components-angular/esm2020/contained-list/contained-list.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ContainedList } from "./contained-list.component";
import { ContainedListItem } from "./contained-list-item.component";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class ContainedListModule {
}
ContainedListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContainedListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ContainedListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ContainedListModule, declarations: [ContainedList, ContainedListItem], imports: [CommonModule, IconModule], exports: [ContainedList, ContainedListItem] });
ContainedListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContainedListModule, imports: [CommonModule, IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContainedListModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ContainedList, ContainedListItem],
                    exports: [ContainedList, ContainedListItem],
                    imports: [CommonModule, IconModule]
                }]
        }] });
