// carbon-components-angular/esm2020/timepicker-select/timepicker-select.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { TimePickerSelect } from "./timepicker-select.component";
import { SelectModule } from "carbon-components-angular/select";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class TimePickerSelectModule {
}
TimePickerSelectModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TimePickerSelectModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TimePickerSelectModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: TimePickerSelectModule, declarations: [TimePickerSelect], imports: [SelectModule,
        CommonModule,
        IconModule], exports: [TimePickerSelect] });
TimePickerSelectModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TimePickerSelectModule, imports: [SelectModule,
        CommonModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TimePickerSelectModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        TimePickerSelect
                    ],
                    exports: [
                        TimePickerSelect
                    ],
                    imports: [
                        SelectModule,
                        CommonModule,
                        IconModule
                    ]
                }]
        }] });
