// carbon-components-angular/esm2020/ui-shell/panel/panel.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { I18nModule } from "carbon-components-angular/i18n";
import { Panel } from "./panel.component";
import { SwitcherList } from "./switcher-list.component";
import { SwitcherListItem } from "./switcher-list-item.component";
import * as i0 from "@angular/core";
export { Panel, SwitcherList, SwitcherListItem };
export class PanelModule {
}
PanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: PanelModule, declarations: [Panel,
        SwitcherList,
        SwitcherListItem], imports: [CommonModule, I18nModule], exports: [Panel,
        SwitcherList,
        SwitcherListItem] });
PanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PanelModule, imports: [CommonModule, I18nModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PanelModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Panel,
                        SwitcherList,
                        SwitcherListItem
                    ],
                    imports: [CommonModule, I18nModule],
                    exports: [
                        Panel,
                        SwitcherList,
                        SwitcherListItem
                    ]
                }]
        }] });
