// carbon-components-angular/esm2020/ui-shell/panel/panel.component.mjs
import { Component, Input } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
/**
 * `Panel` is a component that can be used to display content on the right side of the screen.
 * `Panel`s are generally activated by and linked to a `HeaderAction` component.
 */
export class Panel {
    constructor() {
        /**
         * Controls the visibility of the panel
         */
        this.expanded = false;
    }
}
Panel.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: Panel, deps: [], target: i0.ɵɵFactoryTarget.Component });
Panel.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: Panel, selector: "cds-panel, ibm-panel", inputs: { expanded: "expanded", ariaLabel: "ariaLabel" }, ngImport: i0, template: `
		<div
			class="cds--header-panel"
			[attr.aria-label]="ariaLabel"
			[ngClass]="{
				'cds--header-panel--expanded': expanded
			}">
			<ng-content></ng-content>
		</div>
	`, isInline: true, dependencies: [{ kind: "directive", type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: Panel, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-panel, ibm-panel",
                    template: `
		<div
			class="cds--header-panel"
			[attr.aria-label]="ariaLabel"
			[ngClass]="{
				'cds--header-panel--expanded': expanded
			}">
			<ng-content></ng-content>
		</div>
	`
                }]
        }], propDecorators: { expanded: [{
                type: Input
            }], ariaLabel: [{
                type: Input
            }] } });
