// carbon-components-angular/esm2020/tiles/tiles.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { Tile } from "./tile.component";
import { ClickableTile } from "./clickable-tile.component";
import { ExpandableTile } from "./expandable-tile.component";
import { ExpandableTileAboveFoldDirective } from "./expandable-tile-above.directive";
import { ExpandableTileBelowFoldDirective } from "./expandable-tile-below.directive";
import { SelectionTile } from "./selection-tile.component";
import { TileGroup } from "./tile-group.component";
import { I18nModule } from "carbon-components-angular/i18n";
import { IconModule } from "carbon-components-angular/icon";
import { LinkModule } from "carbon-components-angular/link";
import * as i0 from "@angular/core";
export class TilesModule {
}
TilesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TilesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TilesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: TilesModule, declarations: [Tile,
        ClickableTile,
        ExpandableTileAboveFoldDirective,
        ExpandableTileBelowFoldDirective,
        ExpandableTile,
        SelectionTile,
        TileGroup], imports: [CommonModule,
        I18nModule,
        IconModule,
        LinkModule], exports: [Tile,
        ClickableTile,
        ExpandableTileAboveFoldDirective,
        ExpandableTileBelowFoldDirective,
        ExpandableTile,
        SelectionTile,
        TileGroup] });
TilesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TilesModule, imports: [CommonModule,
        I18nModule,
        IconModule,
        LinkModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TilesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Tile,
                        ClickableTile,
                        ExpandableTileAboveFoldDirective,
                        ExpandableTileBelowFoldDirective,
                        ExpandableTile,
                        SelectionTile,
                        TileGroup
                    ],
                    exports: [
                        Tile,
                        ClickableTile,
                        ExpandableTileAboveFoldDirective,
                        ExpandableTileBelowFoldDirective,
                        ExpandableTile,
                        SelectionTile,
                        TileGroup
                    ],
                    imports: [
                        CommonModule,
                        I18nModule,
                        IconModule,
                        LinkModule
                    ]
                }]
        }] });
