// carbon-components-angular/esm2020/skeleton/skeleton.module.mjs
import { SkeletonPlaceholder } from "./skeleton-placeholder.component";
import { SkeletonText } from "./skeleton-text.component";
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import * as i0 from "@angular/core";
export class SkeletonModule {
}
SkeletonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SkeletonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SkeletonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: SkeletonModule, declarations: [SkeletonPlaceholder,
        SkeletonText], imports: [CommonModule], exports: [SkeletonPlaceholder,
        SkeletonText] });
SkeletonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SkeletonModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SkeletonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        SkeletonPlaceholder,
                        SkeletonText
                    ],
                    exports: [
                        SkeletonPlaceholder,
                        SkeletonText
                    ],
                    imports: [
                        CommonModule
                    ]
                }]
        }] });
