// carbon-components-angular/esm2020/pagination/pagination.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { FormsModule } from "@angular/forms";
import { Pagination } from "./pagination.component";
import { I18nModule } from "carbon-components-angular/i18n";
import { ExperimentalModule } from "carbon-components-angular/experimental";
import { PaginationNav } from "./pagination-nav/pagination-nav.component";
import { PaginationNavItem } from "./pagination-nav/pagination-item.component";
import { PaginationOverflow } from "./pagination-nav/pagination-overflow.component";
import { IconModule } from "carbon-components-angular/icon";
import { ButtonModule } from "carbon-components-angular/forms";
import * as i0 from "@angular/core";
export class PaginationModule {
}
PaginationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PaginationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PaginationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: PaginationModule, declarations: [Pagination,
        PaginationNav,
        PaginationNavItem,
        PaginationOverflow], imports: [CommonModule,
        FormsModule,
        I18nModule,
        ExperimentalModule,
        IconModule,
        ButtonModule], exports: [Pagination,
        PaginationNav,
        PaginationNavItem,
        PaginationOverflow] });
PaginationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PaginationModule, imports: [CommonModule,
        FormsModule,
        I18nModule,
        ExperimentalModule,
        IconModule,
        ButtonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: PaginationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Pagination,
                        PaginationNav,
                        PaginationNavItem,
                        PaginationOverflow
                    ],
                    exports: [
                        Pagination,
                        PaginationNav,
                        PaginationNavItem,
                        PaginationOverflow
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        I18nModule,
                        ExperimentalModule,
                        IconModule,
                        ButtonModule
                    ]
                }]
        }] });
