// carbon-components-angular/esm2020/code-snippet/code-snippet.module.mjs
// modules
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ButtonModule } from "carbon-components-angular/button";
import { IconModule } from "carbon-components-angular/icon";
import { I18nModule } from "carbon-components-angular/i18n";
import { UtilsModule } from "carbon-components-angular/utils";
// imports
import { CodeSnippet } from "./code-snippet.component";
import * as i0 from "@angular/core";
export class CodeSnippetModule {
}
CodeSnippetModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: CodeSnippetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CodeSnippetModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: CodeSnippetModule, declarations: [CodeSnippet], imports: [CommonModule,
        ButtonModule,
        I18nModule,
        UtilsModule,
        IconModule], exports: [CodeSnippet] });
CodeSnippetModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: CodeSnippetModule, imports: [CommonModule,
        ButtonModule,
        I18nModule,
        UtilsModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: CodeSnippetModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        CodeSnippet
                    ],
                    exports: [
                        CodeSnippet
                    ],
                    imports: [
                        CommonModule,
                        ButtonModule,
                        I18nModule,
                        UtilsModule,
                        IconModule
                    ]
                }]
        }] });
