// carbon-components-angular/esm2020/table/expanded-row-hover.directive.mjs
import { Directive, HostListener } from "@angular/core";
import * as i0 from "@angular/core";
export class ExpandedRowHover {
    addHoverClass(event) {
        event.target.previousElementSibling.classList.add("cds--expandable-row--hover");
    }
    removeHoverClass(event) {
        event.target.previousElementSibling.classList.remove("cds--expandable-row--hover");
    }
}
ExpandedRowHover.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ExpandedRowHover, deps: [], target: i0.ɵɵFactoryTarget.Directive });
ExpandedRowHover.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: ExpandedRowHover, selector: "[cdsExpandedRowHover], [ibmExpandedRowHover]", host: { listeners: { "mouseenter": "addHoverClass($event)", "mouseleave": "removeHoverClass($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ExpandedRowHover, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsExpandedRowHover], [ibmExpandedRowHover]"
                }]
        }], propDecorators: { addHoverClass: [{
                type: HostListener,
                args: ["mouseenter", ["$event"]]
            }], removeHoverClass: [{
                type: HostListener,
                args: ["mouseleave", ["$event"]]
            }] } });
