// carbon-components-angular/esm2020/table/table.directive.mjs
import { Directive, HostBinding, Input } from "@angular/core";
import * as i0 from "@angular/core";
export class TableDirective {
    constructor() {
        this.sortable = true;
        this.noBorder = true;
        this.striped = false;
        this.skeleton = false;
        /**
         * Size of the table rows.
         */
        this.size = "md";
        this.tableClass = true;
    }
    // Bind table size class
    get extraSmallSize() { return this.size === "xs"; }
    get smallSize() { return this.size === "sm"; }
    get mediumSize() { return this.size === "md"; }
    get LargeSize() { return this.size === "lg"; }
    get extraLargeSize() { return this.size === "xl"; }
}
TableDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TableDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: TableDirective, selector: "[cdsTable], [ibmTable]", inputs: { sortable: "sortable", noBorder: "noBorder", striped: "striped", skeleton: "skeleton", size: "size" }, host: { properties: { "class.cds--data-table--sort": "this.sortable", "class.cds--data-table--no-border": "this.noBorder", "class.cds--data-table--zebra": "this.striped", "class.cds--skeleton": "this.skeleton", "class.cds--data-table": "this.tableClass", "class.cds--data-table--xs": "this.extraSmallSize", "class.cds--data-table--sm": "this.smallSize", "class.cds--data-table--md": "this.mediumSize", "class.cds--data-table--lg": "this.LargeSize", "class.cds--data-table--xl": "this.extraLargeSize" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsTable], [ibmTable]"
                }]
        }], propDecorators: { sortable: [{
                type: Input
            }, {
                type: HostBinding,
                args: ["class.cds--data-table--sort"]
            }], noBorder: [{
                type: Input
            }, {
                type: HostBinding,
                args: ["class.cds--data-table--no-border"]
            }], striped: [{
                type: Input
            }, {
                type: HostBinding,
                args: ["class.cds--data-table--zebra"]
            }], skeleton: [{
                type: Input
            }, {
                type: HostBinding,
                args: ["class.cds--skeleton"]
            }], size: [{
                type: Input
            }], tableClass: [{
                type: HostBinding,
                args: ["class.cds--data-table"]
            }], extraSmallSize: [{
                type: HostBinding,
                args: ["class.cds--data-table--xs"]
            }], smallSize: [{
                type: HostBinding,
                args: ["class.cds--data-table--sm"]
            }], mediumSize: [{
                type: HostBinding,
                args: ["class.cds--data-table--md"]
            }], LargeSize: [{
                type: HostBinding,
                args: ["class.cds--data-table--lg"]
            }], extraLargeSize: [{
                type: HostBinding,
                args: ["class.cds--data-table--xl"]
            }] } });
