// carbon-components-angular/esm2020/slider/slider.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { Slider } from "./slider.component";
import { UtilsModule } from "carbon-components-angular/utils";
import * as i0 from "@angular/core";
export class SliderModule {
}
SliderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SliderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SliderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: SliderModule, declarations: [Slider], imports: [CommonModule,
        UtilsModule], exports: [Slider] });
SliderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SliderModule, imports: [CommonModule,
        UtilsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SliderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [Slider],
                    exports: [Slider],
                    imports: [
                        CommonModule,
                        UtilsModule
                    ]
                }]
        }] });
