// carbon-components-angular/esm2020/tooltip/tooltip.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { PopoverModule } from "carbon-components-angular/popover";
import { Tooltip } from "./tooltip.component";
import { TooltipDefinition } from "./definition-tooptip.component";
import * as i0 from "@angular/core";
export class TooltipModule {
}
TooltipModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TooltipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TooltipModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: TooltipModule, declarations: [Tooltip,
        TooltipDefinition], imports: [CommonModule, PopoverModule], exports: [Tooltip,
        TooltipDefinition] });
TooltipModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TooltipModule, imports: [CommonModule, PopoverModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TooltipModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Tooltip,
                        TooltipDefinition
                    ],
                    exports: [
                        Tooltip,
                        TooltipDefinition
                    ],
                    imports: [CommonModule, PopoverModule]
                }]
        }] });
