// carbon-components-angular/esm2020/file-uploader/file-uploader.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { FileUploader } from "./file-uploader.component";
import { FileComponent } from "./file.component";
import { ButtonModule } from "carbon-components-angular/button";
import { LoadingModule } from "carbon-components-angular/loading";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class FileUploaderModule {
}
FileUploaderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: FileUploaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FileUploaderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: FileUploaderModule, declarations: [FileUploader, FileComponent], imports: [CommonModule,
        ButtonModule,
        LoadingModule,
        IconModule], exports: [FileUploader, FileComponent] });
FileUploaderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: FileUploaderModule, imports: [CommonModule,
        ButtonModule,
        LoadingModule,
        IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: FileUploaderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [FileUploader, FileComponent],
                    exports: [FileUploader, FileComponent],
                    imports: [
                        CommonModule,
                        ButtonModule,
                        LoadingModule,
                        IconModule
                    ]
                }]
        }] });
