// carbon-components-angular/esm2020/grid/grid.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ColumnDirective } from "./column.directive";
import { RowDirective } from "./row.directive";
import { GridDirective } from "./grid.directive";
import { GridService } from "./grid.service";
import * as i0 from "@angular/core";
export class GridModule {
}
GridModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: GridModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
GridModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: GridModule, declarations: [ColumnDirective,
        GridDirective,
        RowDirective], imports: [CommonModule], exports: [ColumnDirective,
        GridDirective,
        RowDirective] });
GridModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: GridModule, providers: [GridService], imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: GridModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ColumnDirective,
                        GridDirective,
                        RowDirective
                    ],
                    exports: [
                        ColumnDirective,
                        GridDirective,
                        RowDirective
                    ],
                    providers: [GridService],
                    imports: [CommonModule]
                }]
        }] });
