// carbon-components-angular/esm2020/grid/grid.service.mjs
import { Injectable } from "@angular/core";
import { BehaviorSubject } from "rxjs";
import * as i0 from "@angular/core";
export class GridService {
    constructor() {
        this.gridSubject = new BehaviorSubject(false);
        this.cssGridEnabled = false;
        this.gridObservable = this.gridSubject.asObservable();
    }
    /**
     * Ping all subscribers to update to use Css Grid
     * @param enableCssGrid
     */
    updateGridType(enableCssGrid) {
        if (this.cssGridEnabled === enableCssGrid) {
            return;
        }
        this.cssGridEnabled = true;
        this.gridSubject.next(enableCssGrid);
    }
}
GridService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: GridService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
GridService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: GridService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: GridService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return []; } });
