// carbon-components-angular/esm2020/grid/row.directive.mjs
import { Directive, HostBinding, Input } from "@angular/core";
import * as i0 from "@angular/core";
export class RowDirective {
    constructor() {
        this.baseClass = true;
        this.condensed = false;
        this.narrow = false;
    }
}
RowDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: RowDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
RowDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: RowDirective, selector: "[cdsRow], [ibmRow]", inputs: { condensed: "condensed", narrow: "narrow" }, host: { properties: { "class.cds--row": "this.baseClass", "class.cds--row--condensed": "this.condensed", "class.cds--row--narrow": "this.narrow" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: RowDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsRow], [ibmRow]"
                }]
        }], propDecorators: { baseClass: [{
                type: HostBinding,
                args: ["class.cds--row"]
            }], condensed: [{
                type: HostBinding,
                args: ["class.cds--row--condensed"]
            }, {
                type: Input
            }], narrow: [{
                type: HostBinding,
                args: ["class.cds--row--narrow"]
            }, {
                type: Input
            }] } });
