// carbon-components-angular/esm2020/combobox/combobox.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ComboBox } from "./combobox.component";
import { DropdownModule, DropdownService } from "carbon-components-angular/dropdown";
import { I18nModule } from "carbon-components-angular/i18n";
import { UtilsModule } from "carbon-components-angular/utils";
import { IconModule } from "carbon-components-angular/icon";
import * as i0 from "@angular/core";
export class ComboBoxModule {
}
ComboBoxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ComboBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ComboBoxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ComboBoxModule, declarations: [ComboBox], imports: [CommonModule,
        DropdownModule,
        I18nModule,
        UtilsModule,
        IconModule], exports: [ComboBox,
        DropdownModule] });
ComboBoxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ComboBoxModule, providers: [DropdownService], imports: [CommonModule,
        DropdownModule,
        I18nModule,
        UtilsModule,
        IconModule, DropdownModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ComboBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ComboBox
                    ],
                    exports: [
                        ComboBox,
                        DropdownModule
                    ],
                    imports: [
                        CommonModule,
                        DropdownModule,
                        I18nModule,
                        UtilsModule,
                        IconModule
                    ],
                    providers: [DropdownService]
                }]
        }] });
