// carbon-components-angular/esm2020/placeholder/placeholder.component.mjs
import { Component, ViewContainerRef, ViewChild, Input } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "./placeholder.service";
/**
 * Using a modal, dialog (Tooltip, OverflowMenu), or any other component that draws out of the normal page flow
 * in your application *requires* this component (`cds-placeholder`).
 * It would generally be placed near the end of your root app component template
 * (app.component.ts or app.component.html) as:
 *
 * ```
 * <cds-placeholder></cds-placeholder>
 * ```
 */
export class Placeholder {
    /**
     * Creates an instance of `Placeholder`.
     */
    constructor(placeholderService) {
        this.placeholderService = placeholderService;
    }
    /**
     * Registers the components view with `PlaceholderService`
     */
    ngOnInit() {
        // TODO use `id` to register with the placeholderService
        this.placeholderService.registerViewContainerRef(this.viewContainerRef);
    }
}
Placeholder.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: Placeholder, deps: [{ token: i1.PlaceholderService }], target: i0.ɵɵFactoryTarget.Component });
Placeholder.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: Placeholder, selector: "cds-placeholder, ibm-placeholder", inputs: { id: "id" }, viewQueries: [{ propertyName: "viewContainerRef", first: true, predicate: ["placeholder"], descendants: true, read: ViewContainerRef, static: true }], ngImport: i0, template: `<div #placeholder></div>`, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: Placeholder, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-placeholder, ibm-placeholder",
                    template: `<div #placeholder></div>`
                }]
        }], ctorParameters: function () { return [{ type: i1.PlaceholderService }]; }, propDecorators: { id: [{
                type: Input
            }], viewContainerRef: [{
                type: ViewChild,
                args: ["placeholder", { read: ViewContainerRef, static: true }]
            }] } });
