// carbon-components-angular/esm2020/list/list.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ListItemDirective } from "./list-item.directive";
import { List } from "./list.directive";
import * as i0 from "@angular/core";
export class ListModule {
}
ListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ListModule, declarations: [ListItemDirective,
        List], imports: [CommonModule], exports: [ListItemDirective,
        List] });
ListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ListModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ListModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ListItemDirective,
                        List
                    ],
                    exports: [
                        ListItemDirective,
                        List
                    ],
                    imports: [CommonModule]
                }]
        }] });
