// carbon-components-angular/esm2020/layout/layout.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { StackDirective } from "./stack.directive";
import * as i0 from "@angular/core";
export class LayoutModule {
}
LayoutModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LayoutModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LayoutModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: LayoutModule, declarations: [StackDirective], imports: [CommonModule], exports: [StackDirective] });
LayoutModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LayoutModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LayoutModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [StackDirective],
                    exports: [StackDirective],
                    imports: [CommonModule]
                }]
        }] });
