// carbon-components-angular/esm2020/timepicker/timepicker.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { TimePicker } from "./timepicker.component";
import { TimePickerSelectModule } from "carbon-components-angular/timepicker-select";
import * as i0 from "@angular/core";
export class TimePickerModule {
}
TimePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TimePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TimePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: TimePickerModule, declarations: [TimePicker], imports: [TimePickerSelectModule,
        CommonModule], exports: [TimePicker] });
TimePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TimePickerModule, imports: [TimePickerSelectModule,
        CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TimePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        TimePicker
                    ],
                    exports: [
                        TimePicker
                    ],
                    imports: [
                        TimePickerSelectModule,
                        CommonModule
                    ]
                }]
        }] });
