// carbon-components-angular/esm2020/button/button.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { Button } from "./button.directive";
import { ButtonSet } from "./button-set.component";
import { BaseIconButton } from "./base-icon-button.component";
import { IconButton } from "./icon-button.component";
import { TooltipModule } from "carbon-components-angular/tooltip";
import * as i0 from "@angular/core";
export class ButtonModule {
}
ButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ButtonModule, declarations: [Button,
        ButtonSet,
        BaseIconButton,
        IconButton], imports: [CommonModule, TooltipModule], exports: [Button,
        ButtonSet,
        IconButton] });
ButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ButtonModule, imports: [CommonModule, TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Button,
                        ButtonSet,
                        BaseIconButton,
                        IconButton
                    ],
                    exports: [
                        Button,
                        ButtonSet,
                        IconButton
                    ],
                    imports: [CommonModule, TooltipModule]
                }]
        }] });
