// carbon-components-angular/esm2020/content-switcher/content-switcher-option.directive.mjs
import { Directive, HostBinding, Input, HostListener, Output, EventEmitter } from "@angular/core";
import * as i0 from "@angular/core";
export class ContentSwitcherOption {
    constructor(renderer, hostElement) {
        this.renderer = renderer;
        this.hostElement = hostElement;
        /**
         * Internal name for the option.
         * Should be something that identifies the option to the application.
         * Accessible from the `ContentSwitcher` `selected` emitter
         */
        this.name = "option";
        /**
         * Emits when the option is selected.
         */
        this.selected = new EventEmitter();
        this.onClick = new EventEmitter();
        this.onFocus = new EventEmitter();
        this.switcherClass = "cds--content-switcher-btn";
        this.selectedClass = false;
        this.role = "tab";
        this.ariaSelected = false;
        this.tabindex = "-1";
        this._active = false;
    }
    /**
     * Used to activate the option. Only one option may be `active` at a time
     */
    set active(value) {
        this._active = value;
        this.selectedClass = value;
        this.ariaSelected = value;
        this.tabindex = value ? "0" : "-1";
    }
    get active() {
        return this._active;
    }
    hostClick(event) {
        this.onClick.emit(event);
        // skip setting and emitting if the option is already active
        if (this.active) {
            return;
        }
        this.active = true;
        this.selected.emit(true);
    }
    doFocus(event) {
        this.onFocus.emit(event);
        // skip setting and emitting if the option is already active
        if (this.active) {
            return;
        }
        this.active = true;
        this.selected.emit(true);
    }
    /*
    * encapsulating the content in a span with cds--content-switcher__label class
    * to mimic what is done in the react version
    */
    ngOnInit() {
        const hostNativeElement = this.hostElement.nativeElement;
        const spanWrapper = this.renderer.createElement("span");
        this.renderer.addClass(spanWrapper, "cds--content-switcher__label");
        const hostChildren = [];
        hostNativeElement.childNodes.forEach(node => hostChildren.push(node));
        hostChildren.forEach(node => {
            this.renderer.removeChild(hostNativeElement, node);
            this.renderer.appendChild(spanWrapper, node);
        });
        this.renderer.appendChild(hostNativeElement, spanWrapper);
    }
}
ContentSwitcherOption.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContentSwitcherOption, deps: [{ token: i0.Renderer2 }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
ContentSwitcherOption.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: ContentSwitcherOption, selector: "[cdsContentOption], [ibmContentOption]", inputs: { active: "active", name: "name" }, outputs: { selected: "selected", onClick: "onClick", onFocus: "onFocus" }, host: { listeners: { "click": "hostClick($event)", "focus": "doFocus($event)" }, properties: { "class": "this.switcherClass", "class.cds--content-switcher--selected": "this.selectedClass", "attr.role": "this.role", "attr.aria-selected": "this.ariaSelected", "attr.tabIndex": "this.tabindex" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContentSwitcherOption, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsContentOption], [ibmContentOption]"
                }]
        }], ctorParameters: function () { return [{ type: i0.Renderer2 }, { type: i0.ElementRef }]; }, propDecorators: { active: [{
                type: Input
            }], name: [{
                type: Input
            }], selected: [{
                type: Output
            }], onClick: [{
                type: Output
            }], onFocus: [{
                type: Output
            }], switcherClass: [{
                type: HostBinding,
                args: ["class"]
            }], selectedClass: [{
                type: HostBinding,
                args: ["class.cds--content-switcher--selected"]
            }], role: [{
                type: HostBinding,
                args: ["attr.role"]
            }], ariaSelected: [{
                type: HostBinding,
                args: ["attr.aria-selected"]
            }], tabindex: [{
                type: HostBinding,
                args: ["attr.tabIndex"]
            }], hostClick: [{
                type: HostListener,
                args: ["click", ["$event"]]
            }], doFocus: [{
                type: HostListener,
                args: ["focus", ["$event"]]
            }] } });
