// carbon-components-angular/esm2020/content-switcher/content-switcher.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ContentSwitcher } from "./content-switcher.component";
import { ContentSwitcherOption } from "./content-switcher-option.directive";
import * as i0 from "@angular/core";
export class ContentSwitcherModule {
}
ContentSwitcherModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContentSwitcherModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ContentSwitcherModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ContentSwitcherModule, declarations: [ContentSwitcher,
        ContentSwitcherOption], imports: [CommonModule], exports: [ContentSwitcher,
        ContentSwitcherOption] });
ContentSwitcherModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContentSwitcherModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContentSwitcherModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ContentSwitcher,
                        ContentSwitcherOption
                    ],
                    exports: [
                        ContentSwitcher,
                        ContentSwitcherOption
                    ],
                    imports: [CommonModule]
                }]
        }] });
