// carbon-components-angular/esm2020/context-menu/context-menu.module.mjs
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { IconModule } from "carbon-components-angular/icon";
import { ContextMenuDividerComponent } from "./context-menu-divider.component";
import { ContextMenuGroupComponent } from "./context-menu-group.component";
import { ContextMenuItemComponent } from "./context-menu-item.component";
import { ContextMenuComponent } from "./context-menu.component";
import * as i0 from "@angular/core";
export class ContextMenuModule {
}
ContextMenuModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContextMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ContextMenuModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ContextMenuModule, declarations: [ContextMenuDividerComponent,
        ContextMenuGroupComponent,
        ContextMenuItemComponent,
        ContextMenuComponent], imports: [CommonModule, IconModule], exports: [ContextMenuDividerComponent,
        ContextMenuGroupComponent,
        ContextMenuItemComponent,
        ContextMenuComponent] });
ContextMenuModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContextMenuModule, imports: [CommonModule, IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContextMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ContextMenuDividerComponent,
                        ContextMenuGroupComponent,
                        ContextMenuItemComponent,
                        ContextMenuComponent
                    ],
                    exports: [
                        ContextMenuDividerComponent,
                        ContextMenuGroupComponent,
                        ContextMenuItemComponent,
                        ContextMenuComponent
                    ],
                    imports: [CommonModule, IconModule]
                }]
        }] });
