// carbon-components-angular/esm2020/context-menu/context-menu-divider.component.mjs
import { Component, HostBinding } from "@angular/core";
import * as i0 from "@angular/core";
export class ContextMenuDividerComponent {
    constructor() {
        this.dividerClass = true;
        this.role = "separator";
    }
}
ContextMenuDividerComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContextMenuDividerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
ContextMenuDividerComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: ContextMenuDividerComponent, selector: "cds-context-menu-divider, ibm-context-menu-divider", host: { properties: { "class.cds--menu-item-divider": "this.dividerClass", "attr.role": "this.role" } }, ngImport: i0, template: "", isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ContextMenuDividerComponent, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-context-menu-divider, ibm-context-menu-divider",
                    template: ""
                }]
        }], propDecorators: { dividerClass: [{
                type: HostBinding,
                args: ["class.cds--menu-item-divider"]
            }], role: [{
                type: HostBinding,
                args: ["attr.role"]
            }] } });
