// carbon-components-angular/esm2020/utils/element.service.mjs
import { Injectable } from "@angular/core";
import { from } from "rxjs";
import { map } from "rxjs/operators";
import { getScrollableParents, isVisibleInContainer } from "./scroll";
import * as i0 from "@angular/core";
import * as i1 from "./animation-frame.service";
export class ElementService {
    constructor(singleton) {
        this.singleton = singleton;
        this.tick = from(this.singleton.tick);
    }
    visibility(target, parentElement = target) {
        const scrollableParents = getScrollableParents(parentElement);
        return this.tick.pipe(map(() => {
            for (const parent of scrollableParents) {
                if (!isVisibleInContainer(target, parent)) {
                    return {
                        visible: false
                    };
                }
            }
            return {
                visible: true
            };
        }));
    }
}
ElementService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ElementService, deps: [{ token: i1.AnimationFrameServiceSingleton }], target: i0.ɵɵFactoryTarget.Injectable });
ElementService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ElementService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ElementService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.AnimationFrameServiceSingleton }]; } });
