// carbon-components-angular/esm2020/tabs/tab-header.directive.mjs
import { Directive, Input, EventEmitter, Output, HostBinding, HostListener } from "@angular/core";
import * as i0 from "@angular/core";
export class TabHeader {
    constructor(host) {
        this.host = host;
        /**
         * Indicates whether the `Tab` is active/selected.
         * Determines whether it's `TabPanel` is rendered.
         */
        this.active = false;
        /**
         * Indicates whether or not the `Tab` item is disabled.
         */
        this.disabled = false;
        this.type = "button";
        this.ariaSelected = this.active;
        this.ariaDisabled = this.disabled;
        this.navItem = true;
        this.navLink = true;
        /**
         * Value 'selected' to be emitted after a new `Tab` is selected.
         */
        this.selected = new EventEmitter();
        this._cacheActive = false;
    }
    get tabIndex() {
        return this.active ? 0 : -1;
    }
    get isSelected() {
        return this.active;
    }
    get isDisabled() {
        return this.disabled;
    }
    /**
     * Set to 'true' to have pane reference cached and not reloaded on tab switching.
     */
    set cacheActive(shouldCache) {
        this._cacheActive = shouldCache;
        // Updates the pane references associated with the tab header when cache active is changed.
        if (this.paneReference) {
            this.paneReference.cacheActive = this.cacheActive;
        }
    }
    set paneTabIndex(tabIndex) {
        if (this.paneReference) {
            this.paneReference.tabIndex = tabIndex;
        }
    }
    get cacheActive() {
        return this._cacheActive;
    }
    onClick() {
        this.selectTab();
    }
    ngAfterViewInit() {
        setTimeout(() => {
            this.title = this.title ? this.title : this.host.nativeElement.textContent;
        });
    }
    selectTab() {
        this.focus();
        if (!this.disabled) {
            this.selected.emit();
            this.active = true;
            if (this.paneReference) {
                this.paneReference.active = true;
            }
        }
    }
    focus() {
        this.host.nativeElement.focus();
    }
}
TabHeader.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TabHeader, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TabHeader.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: TabHeader, selector: "[cdsTabHeader], [ibmTabHeader]", inputs: { cacheActive: "cacheActive", paneTabIndex: "paneTabIndex", active: "active", disabled: "disabled", paneReference: "paneReference", title: "title" }, outputs: { selected: "selected" }, host: { listeners: { "click": "onClick()" }, properties: { "attr.tabIndex": "this.tabIndex", "class.cds--tabs__nav-item--selected": "this.isSelected", "class.cds--tabs__nav-item--disabled": "this.isDisabled", "attr.type": "this.type", "attr.aria-selected": "this.ariaSelected", "attr.aria-disabled": "this.ariaDisabled", "class.cds--tabs__nav-item": "this.navItem", "class.cds--tabs__nav-link": "this.navLink", "attr.title": "this.title" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TabHeader, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsTabHeader], [ibmTabHeader]"
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { tabIndex: [{
                type: HostBinding,
                args: ["attr.tabIndex"]
            }], isSelected: [{
                type: HostBinding,
                args: ["class.cds--tabs__nav-item--selected"]
            }], isDisabled: [{
                type: HostBinding,
                args: ["class.cds--tabs__nav-item--disabled"]
            }], cacheActive: [{
                type: Input
            }], paneTabIndex: [{
                type: Input
            }], active: [{
                type: Input
            }], disabled: [{
                type: Input
            }], type: [{
                type: HostBinding,
                args: ["attr.type"]
            }], ariaSelected: [{
                type: HostBinding,
                args: ["attr.aria-selected"]
            }], ariaDisabled: [{
                type: HostBinding,
                args: ["attr.aria-disabled"]
            }], navItem: [{
                type: HostBinding,
                args: ["class.cds--tabs__nav-item"]
            }], navLink: [{
                type: HostBinding,
                args: ["class.cds--tabs__nav-link"]
            }], paneReference: [{
                type: Input
            }], title: [{
                type: HostBinding,
                args: ["attr.title"]
            }, {
                type: Input
            }], selected: [{
                type: Output
            }], onClick: [{
                type: HostListener,
                args: ["click"]
            }] } });
