// carbon-components-angular/esm2020/tabs/tabs.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { UtilsModule } from "carbon-components-angular/utils";
import { I18nModule } from "carbon-components-angular/i18n";
import { TabSkeleton } from "./tab-skeleton.component";
import { BaseTabHeader } from "./base-tab-header.component";
import { Tabs } from "./tabs.component";
import { Tab } from "./tab.component";
import { TabHeader } from "./tab-header.directive";
import { TabHeaders } from "./tab-headers.component";
import { TabHeaderGroup } from "./tab-header-group.component";
import * as i0 from "@angular/core";
export class TabsModule {
}
TabsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TabsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TabsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: TabsModule, declarations: [BaseTabHeader,
        Tabs,
        Tab,
        TabHeader,
        TabHeaders,
        TabHeaderGroup,
        TabSkeleton], imports: [CommonModule,
        UtilsModule,
        I18nModule], exports: [Tabs,
        Tab,
        TabHeader,
        TabHeaders,
        TabHeaderGroup,
        TabSkeleton] });
TabsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TabsModule, imports: [CommonModule,
        UtilsModule,
        I18nModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TabsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        BaseTabHeader,
                        Tabs,
                        Tab,
                        TabHeader,
                        TabHeaders,
                        TabHeaderGroup,
                        TabSkeleton
                    ],
                    exports: [
                        Tabs,
                        Tab,
                        TabHeader,
                        TabHeaders,
                        TabHeaderGroup,
                        TabSkeleton
                    ],
                    imports: [
                        CommonModule,
                        UtilsModule,
                        I18nModule
                    ]
                }]
        }] });
