// carbon-components-angular/esm2020/theme/theme.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ThemeDirective } from "./theme.directive";
import { LayerModule } from "carbon-components-angular/layer";
import * as i0 from "@angular/core";
export class ThemeModule {
}
ThemeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ThemeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ThemeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ThemeModule, declarations: [ThemeDirective], imports: [CommonModule, LayerModule], exports: [ThemeDirective] });
ThemeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ThemeModule, imports: [CommonModule, LayerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ThemeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ThemeDirective],
                    exports: [ThemeDirective],
                    imports: [CommonModule, LayerModule]
                }]
        }] });
