// carbon-components-angular/esm2020/toggletip/toggletip-button.directive.mjs
import { Directive, HostBinding, Input } from "@angular/core";
import * as i0 from "@angular/core";
export class ToggletipButton {
    constructor() {
        this.toggletipButton = true;
        this.toggletipButtonType = "button";
        this.ariaLabel = "Show information";
    }
}
ToggletipButton.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ToggletipButton, deps: [], target: i0.ɵɵFactoryTarget.Directive });
ToggletipButton.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: ToggletipButton, selector: "[cdsToggletipButton], [ibmToggletipButton]", inputs: { ariaLabel: "ariaLabel" }, host: { properties: { "class.cds--toggletip-button": "this.toggletipButton", "attr.type": "this.toggletipButtonType", "attr.aria-label": "this.ariaLabel" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ToggletipButton, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsToggletipButton], [ibmToggletipButton]"
                }]
        }], propDecorators: { toggletipButton: [{
                type: HostBinding,
                args: ["class.cds--toggletip-button"]
            }], toggletipButtonType: [{
                type: HostBinding,
                args: ["attr.type"]
            }], ariaLabel: [{
                type: HostBinding,
                args: ["attr.aria-label"]
            }, {
                type: Input
            }] } });
