// carbon-components-angular/esm2020/toggletip/toggletip.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { PopoverModule } from "carbon-components-angular/popover";
import { ToggletipLabel } from "./toggletip-label.directive";
import { ToggletipAction } from "./toggletip-action.directive";
import { Toggletip } from "./toggletip.component";
import { ToggletipButton } from "./toggletip-button.directive";
import { ToggletipContent } from "./toggletip-content.directive";
import * as i0 from "@angular/core";
export class ToggletipModule {
}
ToggletipModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ToggletipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ToggletipModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ToggletipModule, declarations: [Toggletip,
        ToggletipLabel,
        ToggletipAction,
        ToggletipButton,
        ToggletipContent], imports: [CommonModule, PopoverModule], exports: [Toggletip,
        ToggletipLabel,
        ToggletipAction,
        ToggletipButton,
        ToggletipContent] });
ToggletipModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ToggletipModule, imports: [CommonModule, PopoverModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ToggletipModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Toggletip,
                        ToggletipLabel,
                        ToggletipAction,
                        ToggletipButton,
                        ToggletipContent
                    ],
                    exports: [
                        Toggletip,
                        ToggletipLabel,
                        ToggletipAction,
                        ToggletipButton,
                        ToggletipContent
                    ],
                    imports: [CommonModule, PopoverModule]
                }]
        }] });
