// carbon-components-angular/esm2020/progress-indicator/progress-indicator.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ProgressIndicator } from "./progress-indicator.component";
import { IconModule } from "carbon-components-angular/icon";
import { I18nModule } from "carbon-components-angular/i18n";
import * as i0 from "@angular/core";
export class ProgressIndicatorModule {
}
ProgressIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ProgressIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ProgressIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: ProgressIndicatorModule, declarations: [ProgressIndicator], imports: [CommonModule,
        IconModule,
        I18nModule], exports: [ProgressIndicator] });
ProgressIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ProgressIndicatorModule, imports: [CommonModule,
        IconModule,
        I18nModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ProgressIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ProgressIndicator
                    ],
                    exports: [
                        ProgressIndicator
                    ],
                    imports: [
                        CommonModule,
                        IconModule,
                        I18nModule
                    ]
                }]
        }] });
